File: Ejercicio 3/main.py
from movil import Movil
from movilAutonomo import MovilAutonomo
from random import randint
from time import time

print('Ingresar dimensiones de la grilla:')
w = int(input('w = '))
h = int(input('h = '))
print('Elija la dirección de movimiento con:')
print('1. ADELANTE')
print('2. DERECHA')
print('3. ATRAS')
print('4. IZQUIERDA')
print('5. TERMINAR')
print('0. SALIR')
distancia = 0
distancia1 = 0
orden = []
movil = []
print('Ingresar la cantidad de moviles autónomos:')
N = int(input('N = '))

#MOVIL MANUAL
movil0 = Movil('MOVIL_MANUAL')
print(movil0.nombre)
i0 = randint(0, w)
j0 = randint(0, h)
print('La posición actual es: (',i0,',',j0,')')
i = i0
j = j0
print('Insertar secuencia de operaciones:')
b = True
while (b == True):
    op_mover = int(input())
    if (op_mover == 1):
        distancia = distancia + 1
        orden += ['ADELANTE']
        j = movil0.getY(j0, op_mover)
        j0 = j
    if (op_mover == 2):
        distancia = distancia + 1
        orden += ['DERECHA']
        i = movil0.getX(i0, op_mover)
        i0 = i
    if (op_mover == 3):
        distancia = distancia + 1
        orden += ['ATRAS']
        j = movil0.getY(j0, op_mover)
        j0 = j
    if (op_mover == 4):
        distancia = distancia + 1
        orden += ['IZQUIERDA']
        i = movil0.getX(i0, op_mover)
        i0 = i
    if (op_mover == 5):
        orden += ['TERMINAR']
        b = False
    if (op_mover == 0):
        orden += ['SALIR']
        b = False
    if (i > w or i < 0 or j > h or j < 0):
        print('Se sale de la grilla.')
        b = False
print('Nueva posición: (',i,',',j,')')
print('Secuencia de órdenes:')
print(orden)
print('Distancia recorrida:', distancia, '[m]')

orden[:] = []
#MOVILES AUTÓNOMOS
movil1 = MovilAutonomo('MOVIL_AUTONOMO')
print(movil1.nombre)
k = 1
while (k <= N):
    start_time = time()
    movil += [k]
    i01 = randint(0, w)
    j01 = randint(0, h)
    print('La posición actual es: (',i01,',',j01,')')
    i1 = i01
    j1 = j01
    carga = randint(200, 300)
    print('Carga inicial:', carga,'[mAh]')
    indicador = carga
    b = True
    print('Insertar secuencia de operaciones:')
    while (b == True):
        op_mover = int(input())
        if (op_mover == 1):
            distancia1 = distancia1 + 1
            orden += ['ADELANTE']
            j1 = movil1.getY(j01, op_mover)
            j01 = j1
        if (op_mover == 2):
            distancia1 = distancia1 + 1
            orden += ['DERECHA']
            i1 = movil1.getX(i01, op_mover)
            i01 = i1
        if (op_mover == 3):
            distancia1 = distancia1 + 1
            orden += ['ATRAS']
            j1 = movil1.getY(j01, op_mover)
            j01 = j1
        if (op_mover == 4):
            distancia1 = distancia1 + 1
            orden += ['IZQUIERDA']
            i1 = movil1.getX(i01, op_mover)
            i01 = i1
        if (op_mover == 5):
            orden += ['TERMINAR']
            end_time = time()
            b = False
        if (op_mover == 0):
            orden += ['SALIR']
            end_time = time()
            b = False
        if (i1 > w or i1 < 0 or j1 > h or j1 < 0):
            print('Se sale de la grilla.')
            end_time = time()
            b = False
    tiempo = end_time - start_time
    print('Nueva posición: (',i1,',',j1,')')
    print('Secuencia de órdenes:')
    print(orden)
    print('Distancia recorrida:',distancia1,'[m]')
    indicador = movil1.consumo(distancia1, carga)
    c = carga - indicador
    print('Consumió:',c,'[mAh]')
    print('Carga restante:',indicador,'[mAh]')
    print('Tiempo del movimiento:',tiempo,'[s]')
    i01 = 0
    j01 = 0
    start_time = 0
    distancia1 = 0
    orden[:] = []
    k = k + 1
print('Lista de móviles autónomos:')
print(movil)

File: Ejercicio 3/main.cpp
// Modificar el programa que permite gestionar los móviles anterior de modo
// tal que:
// Se puedan crear N móviles autónomos (en lugar de sólo 3)
// Los móviles se almacenan en una lista de objetos
// La secuencia de órdenes recibidas se almacene en otra lista adecuada

#include "movil.h"
#include "movilAutonomo.h"
#include "lista.h"
#include <list>

int main(){
  int w, h;
  double t0, tf;
  cout << "Ingresar dimensiones de la grilla" << endl;
  cout << "w = ";
  cin >> w;
  cout << "h = ";
  cin >> h;
  cout << "Elija la dirección de movimiento con:" << endl
        << "1. ADELANTE" << endl
        << "2. DERECHA" << endl
        << "3. ATRAS" << endl
        << "4. IZQUIERDA" << endl
        << "5. TERMINAR" << endl
        << "0. SALIR" << endl;
  bool b = true;
  int carga, op_mover, ind, N, distancia = 0, i0 = 0, j0 = 0, i = 0, j = 0;
  string orden;
  int q, i01 = 0, j01 = 0, i1 = 0, j1 = 0, distancia1 = 0;

  Movil *m;
  m = new Movil(i, j, i0, j0, w, h, op_mover, orden); //MOVIL MANUAL
  list <MovilAutonomo> m_a; //LISTA DE MOVILES AUTONOMOS
  list <MovilAutonomo>::iterator it;  //ITERADOR DE MOVILES AUTONOMOS
  list <MovilAutonomo> ordenes; //LISTA DE ÓRDENES
  list <MovilAutonomo>::iterator it_orden;  //ITERADOR DE ÓRDENES

  cout << "Ingresar la cantidad de moviles autónomos:" << endl;
  cout << "N = ";
  cin >> N;

  for (q = 0; q < N; q++) {
    m_a.insert(m_a.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
  }

  //MOVIL MANUAL
  cout << "MOVIL_MANUAL" << endl;
  cout << "La posición actual es: (";
  srand(time(NULL));
  i0 = 1 + rand() % w - 1;
  j0 = 1 + rand() % h - 1;
  cout << i0 << "," << j0 << ")" << endl;
  i = i0;
  j = j0;
  cout << "Insertar secuencia de operaciones: " << endl;
  while (b == true){
    cin >> op_mover;
    m -> mover(i0, j0, op_mover);
    orden = m -> getOrden();
    if (op_mover == 1) {
      distancia++;
      ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
      j = m -> getY();
      j0 = j;
    }
    if (op_mover == 2) {
      distancia++;
      ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
      i = m -> getX();
      i0 = i;
    }
    if (op_mover == 3) {
      distancia++;
      ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
      j = m -> getY();
      j0 = j;
    }
    if (op_mover == 4) {
      distancia++;
      ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
      i = m -> getX();
      i0 = i;
    }
    if (op_mover == 5) {
      ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
      b = false;
    }
    if (op_mover == 0) {
      ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
      b = false;
    }
    if (i > w || i < 0 || j > h || j < 0) {
      cout << "Se sale de la grilla." << endl;
      b = false;
    }
  }
  cout << "Nueva posición: (" << i << "," << j << ")" << endl;
  cout << "Secuencia de órdenes: " << endl;
  for (it_orden = ordenes.begin(); it_orden != ordenes.end(); it_orden++){
    cout << it_orden -> toString() << endl;
  }
  cout << "Distancia recorrida: " << distancia << " [m]" << endl;
  //BORRAR LOS OBJETOS DE LA LISTA ORDENES
  ordenes.clear();

  //MOVILES AUTONOMOS
  int n = 1;
  t0 = clock();
  for (it = m_a.begin(); it != m_a.end(); it++) {
    cout << "MOVIL_AUTONOMO Nro. " << n << endl;
    srand(time(NULL));
    i01 = 1 + rand() % w - 1;
    j01 = 1 + rand() % h - 1;
    cout << "La posición actual es: (";
    cout << i01 << "," << j01 << ")" << endl;
    i1 = i01;
    j1 = j01;
    cout << "Carga inicial: ";
    carga = it -> getCarga_inicial();
    cout << carga << "[mAh]" << endl;
    ind = carga;
    b = true;
    cout << "Insertar secuencia de operaciones:" << endl;
    while (b == true) {
      cin >> op_mover;
      it -> mover(i01, j01, op_mover);
      orden = it -> getOrden();
      if (op_mover == 1) {
        distancia1++;
        ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
        j1 = it -> getY();
        j01 = j1;
      }
      if (op_mover == 2) {
        distancia1++;
        ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
        i1 = it -> getX();
        i01 = i1;
      }
      if (op_mover == 3) {
        distancia1++;
        ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
        j1 = it -> getY();
        j01 = j1;
      }
      if (op_mover == 4) {
        distancia1++;
        ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
        i1 = it -> getX();
        i01 = i1;
      }
      if (op_mover == 5) {
        tf = clock();
        ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
        b = false;
      }
      if (op_mover == 0) {
        tf = clock();
        ordenes.insert(ordenes.end(), MovilAutonomo(i1, j1, i01, j01, w, h, op_mover, orden, carga, ind, distancia1));
        b = false;
      }
      if (i1 > w || i1 < 0 || j1 > h || j1 < 0) {
        cout << "Se sale de la grilla." << endl;
        tf = clock();
        b = false;
      }
    }
    double tiempo = (double(tf-t0)/CLOCKS_PER_SEC);
    cout << "Nueva posición: (" << i1 << "," << j1 << ")" << endl;
    cout << "Secuencia de órdenes: " << endl;
    for (it_orden = ordenes.begin(); it_orden != ordenes.end(); it_orden++){
      cout << it_orden -> toString() << endl;
    }
    cout << "Distancia recorrida: " << distancia1 << " [m]" << endl;
    ind = it -> consumo(distancia1);
    int c = carga - ind;
    cout << "Consumió: " << c << " [mAh]" << endl;
    cout << "Carga restante: " << ind << " [mAh]" << endl;
    cout << "Tiempo del movimiento: " << tiempo << "[s]" << endl;
    i01 = 0;
    j01 = 0;
    t0 = 0;
    distancia1 = 0;
    ordenes.clear();
    n++;
  }

  return 0;
}
